### PR TITLE
Document net.minecraft.nbt package

### PR DESCRIPTION
**Pull request by apple502j**
Documents the `net.minecraft.nbt` package with big `package-info.java` file (3 code snippets showing integer handling basics, lists, and compounds) and javadocs to almost all `public` methods.

### CODE REVIEW

File: mappings/net/minecraft/nbt/NbtByte.mapping [hunk 1]
```diff
@@ -1,17 +1,27 @@
 CLASS net/minecraft/class_2481 net/minecraft/nbt/NbtByte
-	COMMENT Represents an NBT byte.
+	COMMENT Represents an NBT byte. Its type is {@value NbtElement#BYTE_TYPE}.
```
Comment (liach) [inline mappings/net/minecraft/nbt/NbtByte.mapping:2]: Maybe document that if they are immutable or not, as this is the case for some number tags but not for list or compounds.

File: mappings/net/minecraft/nbt/NbtCompound.mapping [hunk 2]
```diff
@@ -1,5 +1,19 @@
 CLASS net/minecraft/class_2487 net/minecraft/nbt/NbtCompound
-	COMMENT Represents an NBT compound object which holds unordered key-value pairs with distinct case-sensitive string keys.
+	COMMENT Represents an NBT compound object. This mutable object holds unordered key-value pairs
+	COMMENT with distinct case-sensitive string keys. This can effectively be used like a
+	COMMENT {@code HashMap<String, NbtElement>}. Note that this <strong>does not</strong> implement
+	COMMENT {@link java.util.Map}. Its type is {@value NbtElement#COMPOUND_TYPE}. To get the compound
+	COMMENT as a map, use {@link #toMap()}.
+	COMMENT
+	COMMENT There are two ways to use this compound; one is to create NBT instances yourself and use
```
Comment (liach) [inline mappings/net/minecraft/nbt/NbtCompound.mapping]: Use `<p>`, or it's rendered in the web page without paragraph breaks.

File: mappings/net/minecraft/nbt/NbtByte.mapping [hunk 3]
```diff
@@ -1,17 +1,28 @@
...
+	COMMENT Instances are immutable.
 	FIELD field_11498 value B
 	FIELD field_21025 TYPE Lnet/minecraft/class_4614;
 	FIELD field_21026 ZERO Lnet/minecraft/class_2481;
+		COMMENT The NBT byte representing {@code 0}.
+		COMMENT
+		COMMENT @apiNote This is often used to indicate a falsy value.
```
Comment (liach) [inline mappings/net/minecraft/nbt/NbtByte.mapping]:
```suggestion
		COMMENT @apiNote This is often used to indicate a false boolean value.
```

File: mappings/net/minecraft/nbt/NbtByte.mapping [hunk 4]
```diff
@@ -1,17 +1,28 @@
...
 	FIELD field_21027 ONE Lnet/minecraft/class_2481;
+		COMMENT The NBT byte representing {@code 1}.
+		COMMENT
+		COMMENT @apiNote This is often used to indicate a truthy value.
```
Comment (liach) [inline mappings/net/minecraft/nbt/NbtByte.mapping]:
```suggestion
		COMMENT @apiNote This is often used to indicate a true boolean value.
```

File: mappings/net/minecraft/nbt/NbtElement.mapping [hunk 5]
```diff
@@ -59,20 +59,24 @@ CLASS net/minecraft/class_2520 net/minecraft/nbt/NbtElement
 		COMMENT @see NbtCompound#contains(String, int)
 	FIELD field_33264 MAX_DEPTH I
 	METHOD method_10707 copy ()Lnet/minecraft/class_2520;
-		COMMENT Copies this NBT element.
-		COMMENT
-		COMMENT @return the copied element
+		COMMENT {@return the copied NBT element}
```
Comment (liach) [inline mappings/net/minecraft/nbt/NbtElement.mapping]:
```suggestion
		COMMENT {@return a copy of this NBT element}
```
Also, does this copy always perform a copy, or does it return the element itself if it's immutable?

File: mappings/net/minecraft/nbt/NbtHelper.mapping [hunk 6]
```diff
@@ -1,4 +1,5 @@
 CLASS net/minecraft/class_2512 net/minecraft/nbt/NbtHelper
+	COMMENT Helper methds for handling NBT.
```
Comment (liach) [inline mappings/net/minecraft/nbt/NbtHelper.mapping]:
```suggestion
	COMMENT Helper methods for handling NBT.
```

File: mappings/net/minecraft/nbt/NbtHelper.mapping [hunk 7]
```diff
@@ -10,27 +11,65 @@ CLASS net/minecraft/class_2512 net/minecraft/nbt/NbtHelper
...
 	METHOD method_10685 nameValue (Lnet/minecraft/class_2769;Ljava/lang/Comparable;)Ljava/lang/String;
 		ARG 0 property
 		ARG 1 value
 	METHOD method_10686 fromBlockState (Lnet/minecraft/class_2680;)Lnet/minecraft/class_2487;
+		COMMENT {@return the serialized block state}
+		COMMENT
+		COMMENT @see #toBlockState(NbtCompound)
 		ARG 0 state
 	METHOD method_10687 matches (Lnet/minecraft/class_2520;Lnet/minecraft/class_2520;Z)Z
+		COMMENT {@return whether {@code subject} "matches" {@code standard}}
```
Comment (liach) [inline mappings/net/minecraft/nbt/NbtHelper.mapping]: You can just write `whether {@code standard} is a subset of {@code subject}`

File: mappings/net/minecraft/nbt/NbtList.mapping [hunk 8]
```diff
@@ -16,23 +26,47 @@ CLASS net/minecraft/class_2499 net/minecraft/nbt/NbtList
...
+		COMMENT {@return the float at {@code index}, or {@code 0.0f} if the index is out of bounds
+		COMMENT or if this is not a list of floats}
 		ARG 1 index
 	METHOD method_10605 canAdd (Lnet/minecraft/class_2520;)Z
 		ARG 1 element
 	METHOD method_10608 getString (I)Ljava/lang/String;
+		COMMENT {@return the stringified value at {@code index}, or an empty string if the index
+		COMMENT is out of bounds}
+		COMMENT
+		COMMENT <p>Unlike other getters, this works with any types, not just {@link NbtElement}.
```
Comment (liach) [inline mappings/net/minecraft/nbt/NbtList.mapping]:
```suggestion
		COMMENT <p>Unlike other getters, this works with any type, not just {@link NbtString}.
```
Do you mean this?

File: mappings/net/minecraft/nbt/NbtTagSizeTracker.mapping [hunk 9]
```diff
@@ -1,4 +1,6 @@
 CLASS net/minecraft/class_2505 net/minecraft/nbt/NbtTagSizeTracker
+	COMMENT Tracks the size of NBT elements. If the element became too larger than
+	COMMENT {@link #maxBytes}, adding more elements will raise {@link RuntimeException}.
```
Comment (liach) [inline mappings/net/minecraft/nbt/NbtTagSizeTracker.mapping]:
```suggestion
	COMMENT Tracks the size of NBT elements. Throws {@link RuntimeException} if the
	COMMENT tracked element becomes larger than {@link #maxBytes} during addition.
```

File: src/packageDocs/java/net/minecraft/nbt/package-info.java [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+ * This file is free for everyone to use under the Creative Commons Zero license.
+ */
+
+/**
+ * The Named Binary Tag (NBT) data format.
+ *
+ * <p>NBT is a simple data structure format used in Minecraft's in-game data
+ * serialization. Data stored on disks, modified by commands, and transported
+ * via packets are usually in this format. In most cases, NBT is <strong>not
+ * read or written at runtime</strong>; NBT data are instead written to Java
```
Comment (liach) [inline src/packageDocs/java/net/minecraft/nbt/package-info.java]:
```suggestion
 * the runtime form of data</strong>; data are instead written to Java
```

File: src/packageDocs/java/net/minecraft/nbt/package-info.java [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+ *   // (internally represented as a byte)
+ *   compound.putBoolean("Awesome", true);
+ *   boolean awesome = compound.getBoolean("Awesome");
+*    compound.put("awesome", NbtDouble.of(1.0)); // key is case-sensitive
+ * }</pre>
+ *
+ * <h2 id=using-nbt>Using NBT</h2>
+ * <p>As noted before, NBT is a <strong>serialized format, not runtime format</strong>.
+ * Therefore, modifying NBT is usually not the solution for attaching custom data to
+ * objects like entities. For this purpose it is recommended to use third-party APIs
```
Comment (liach) [inline src/packageDocs/java/net/minecraft/nbt/package-info.java:207]: Mention that it's not a solution as the custom NBT will be lost once processed by the game.

**Review by liach — COMMENTED** [review]
For the overview: I wonder if we should add docs on the binary format of nbt, that each file's root tag has a name. This is something that is quite confusing, especially when people compare nbt to jsons.

File: mappings/net/minecraft/nbt/NbtElement.mapping (hunk 5 above)
Comment (apple502j) [reply mappings/net/minecraft/nbt/NbtElement.mapping]: good point (it returns itself, yes)

File: mappings/net/minecraft/nbt/NbtList.mapping (hunk 8 above)
Comment (apple502j) [reply mappings/net/minecraft/nbt/NbtList.mapping]: Whoops

File: mappings/net/minecraft/nbt/NbtElement.mapping (hunk 5 above)
Comment (liach) [reply mappings/net/minecraft/nbt/NbtElement.mapping]: Maybe like ``
```suggestion
		COMMENT {@return an NBT element of equal value that won't change with this element}
```

File: mappings/net/minecraft/nbt/NbtHelper.mapping (hunk 7 above)
Comment (apple502j) [reply mappings/net/minecraft/nbt/NbtHelper.mapping]:
```suggestion
		COMMENT {@return whether {@code standard} is a subset of {@code subject}}
```

File: src/packageDocs/java/net/minecraft/nbt/package-info.java [hunk 12]
```diff
@@ -199,6 +199,7 @@
  *
  * <h2 id=using-nbt>Using NBT</h2>
  * <p>As noted before, NBT is a <strong>serialized format, not runtime format</strong>.
+ * Most custom data stored inside NBT will be lost once the game loads it and saves.
```
Comment (liach) [inline src/packageDocs/java/net/minecraft/nbt/package-info.java]: Can you change "most" to "unrecognized", to be specific?

**Review by liach — APPROVED** [review]

File: mappings/net/minecraft/nbt/NbtCompound.mapping [hunk 13]
```diff
@@ -9,11 +23,26 @@ CLASS net/minecraft/class_2487 net/minecraft/nbt/NbtCompound
 	METHOD equals (Ljava/lang/Object;)Z
 		ARG 1 o
 	METHOD method_10537 getLong (Ljava/lang/String;)J
+		COMMENT {@return the {@code long} assciated with {@code key}, or {@code 0L} if there is no number
```
Comment (enbrain) [inline mappings/net/minecraft/nbt/NbtCompound.mapping]:
```suggestion
		COMMENT {@return the {@code long} associated with {@code key}, or {@code 0L} if there is no number
```
The same applies to other instances.

File: src/packageDocs/java/net/minecraft/nbt/package-info.java [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+ * <p>NbtList is a list of a specific type of NBT elements. Empty lists always have the type
+ * set as {@link NbtEnd}. {@link NbtList#add} throws when the type is incompatible, while
+ * {@link NbtList#addElement} does nothing and returns {@code false} in that case.
+ * There are type-specific getters which can be used to get the value. They return the default
+ * value when type mismatch occurs or when the index is out of bounds. Note that they do not
+ * cast the value at all.
+ *
+ * <pre>{@code
+ *   NbtList list = new NbtList();
+ *   list.addElement(NbtFloat.of(0.5f));
```
Comment (enbrain) [inline src/packageDocs/java/net/minecraft/nbt/package-info.java]:
```suggestion
 *   list.addElement(0, NbtFloat.of(0.5f));
```

File: src/packageDocs/java/net/minecraft/nbt/package-info.java [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+ *   float firstItem = list.getFloat(0); // 0.5f
+ *   float secondItem = list.getFloat(1); // 0.0f (default value)
+ *   double firstItemDouble = list.getDouble(0); // 0.0 (no casting)
+ *   list.addElement(NbtDouble.of(1.0)); // returns false, does nothing
```
Comment (enbrain) [inline src/packageDocs/java/net/minecraft/nbt/package-info.java]:
```suggestion
 *   list.addElement(1, NbtDouble.of(1.0)); // returns false, does nothing
```

File: src/packageDocs/java/net/minecraft/nbt/package-info.java [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+ * object for serialized NBTs. The keys are always strings, while the values can be of
+ * any type, and multiple value types can be mixed in a single compound. The order of items
+ * is not guaranteed. There are generic {@link NbtCompound#put} and {@link NbtCompound#get}
+ * methods, and there also exist type-specific getters and putters. Like the list getters
+ * mentioned above, they return the default value when the key is missing or type mismatch
+ * occurs. However, unlike lists, they attempt to cast numeric values.
+ *
+ * <pre>{@code
+ *   NbtCompound compound = new NbtCompound();
+ *   compound.put("Awesomeness", NbtInt.of(90));
```
Comment (enbrain) [inline src/packageDocs/java/net/minecraft/nbt/package-info.java:186]:
```suggestion
 *   compound.put("Awesomeness", NbtInt.of(90));
 *
```

File: src/packageDocs/java/net/minecraft/nbt/package-info.java [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+ *   // Don't do this! This will crash if the type is incompatible,
+ *   // even if they can be casted!
+ *   NbtInt awesomenessNbt = (NbtInt)compound.get("Awesomeness");
```
Comment (enbrain) [inline src/packageDocs/java/net/minecraft/nbt/package-info.java:190]:
```suggestion
 *   NbtInt awesomenessNbt = (NbtInt)compound.get("Awesomeness");
 *
```

File: src/packageDocs/java/net/minecraft/nbt/package-info.java [hunk 18]
```diff
@@ -0,0 +1,235 @@
...
+ *   compound.putLong("Awesomeness", 100L);
+ *   int awesomeness = compound.getInt("Awesomeness"); // 100 (after casting)
+ *   int evilness = compound.getInt("Evilness"); // 0 (default value)
```
Comment (enbrain) [inline src/packageDocs/java/net/minecraft/nbt/package-info.java:194]:
```suggestion
 *   int evilness = compound.getInt("Evilness"); // 0 (default value)
 *
```

File: src/packageDocs/java/net/minecraft/nbt/package-info.java [hunk 19]
```diff
@@ -0,0 +1,235 @@
...
+ *   // Shortcuts for getting and putting a boolean value
+ *   // (internally represented as a byte)
+ *   compound.putBoolean("Awesome", true);
+ *   boolean awesome = compound.getBoolean("Awesome");
+*    compound.put("awesome", NbtDouble.of(1.0)); // key is case-sensitive
```
Comment (enbrain) [inline src/packageDocs/java/net/minecraft/nbt/package-info.java]:
```suggestion
 *   compound.put("awesome", NbtDouble.of(1.0)); // key is case-sensitive
```

File: mappings/net/minecraft/nbt/NbtCompound.mapping (hunk 13 above)
Comment (apple502j) [reply mappings/net/minecraft/nbt/NbtCompound.mapping]: whoops, guess ill edit on my phone then

**Review by enbrain — APPROVED** [review]

**Pull request merged**